float64 mode
float64 safety_mode
